### ISSUE TITLE
Make sure to properly terminate not fully established connections

### ISSUE THREAD

**Issue by onderkalaci** [citusdata/citus#4689]
The adaptive executor may have decided to establish connections while the execution is in progress. However, by the time the execution finishes, the connections might not have properly established.

In that case, Citus abruptly terminates the connections, and this fills the logs with errors like:
```SQL
 2020-11-16 21:09:09.800 CET [16633] LOG:  could not accept SSL connection: Connection reset by peer
2020-11-16 21:09:09.872 CET [16657] LOG:  could not accept SSL connection: Undefined error: 0
2020-11-16 21:09:09.894 CET [16667] LOG:  could not accept SSL connection: Connection reset by peer
```

An attempt to fix this which added a lot of perf penalty: https://github.com/citusdata/citus/pull/4317

This is valid for all cases. However, for single node/starter plan, the error logs will be on the coordinator can can be very confusing. Especially under high loads, these errors become more visible.

**Comment by onderkalaci** [comment]
I was planning to calculate the current execution's connection establishment / task execution times. And, depending on the current status, decide on whether or not establish a new connection, such as here:  https://github.com/citusdata/citus/compare/connection_establishment_try

 However, I observed that it doesn't play well on busy servers as both of these values could vary a lot depending on how much CPU they get. For example, see the connection establishment times while the CPUs are maxed out on the nodes, they are almost random.

Implementing a fine tuning based on these values ends up with unexpected/unpredictable query execution time differences. That's why I'll look for another approach 

```
select count(*) from test;
WARNING:  Connections to node 10.192.0.4:5432
WARNING:  Connection establishment time for session 93: 107 msecs
WARNING:  Connection establishment time for session 96: 360 msecs
WARNING:  Connection establishment time for session 97: 365 msecs
WARNING:  Connection establishment time for session 105: 616 msecs
WARNING:  Connection establishment time for session 106: 330 msecs
WARNING:  Connection establishment time for session 107: 486 msecs
WARNING:  Connection establishment time for session 120: 345 msecs
WARNING:  Connection establishment time for session 121: 194 msecs
WARNING:  Connection establishment time for session 122: 422 msecs
WARNING:  Connection establishment time for session 123: 632 msecs
WARNING:  Connection establishment time for session 124: 253 msecs
WARNING:  Connections to node 10.192.0.6:5432
WARNING:  Connection establishment time for session 94: 011 msecs
WARNING:  Connection establishment time for session 98: 309 msecs
WARNING:  Connection establishment time for session 99: 321 msecs
WARNING:  Connection establishment time for session 102: 291 msecs
WARNING:  Connection establishment time for session 103: 303 msecs
WARNING:  Connection establishment time for session 104: 231 msecs
WARNING:  Connection establishment time for session 111: 190 msecs
WARNING:  Connection establishment time for session 112: 082 msecs
WARNING:  Connection establishment time for session 113: 129 msecs
WARNING:  Connection establishment time for session 114: 084 msecs
WARNING:  Connection establishment time for session 115: 251 msecs
WARNING:  Connections to node 10.192.0.7:5432
WARNING:  Connection establishment time for session 95: 011 msecs
WARNING:  Connection establishment time for session 100: 371 msecs
WARNING:  Connection establishment time for session 101: 371 msecs
WARNING:  Connection establishment time for session 108: 504 msecs
WARNING:  Connection establishment time for session 109: 319 msecs
WARNING:  Connection establishment time for session 110: 510 msecs
WARNING:  Connection establishment time for session 116: 185 msecs
WARNING:  Connection establishment time for session 117: 144 msecs
WARNING:  Connection establishment time for session 118: 124 msecs
WARNING:  Connection establishment time for session 119: 140 msecs
  count   
----------
 80000008
(1 row)



select count(*) from test;
WARNING:  Connections to node 10.192.0.4:5432
WARNING:  Connection establishment time for session 46: 472 msecs
WARNING:  Connection establishment time for session 49: 250 msecs
WARNING:  Connection establishment time for session 50: 270 msecs
WARNING:  Connection establishment time for session 62: 219 msecs
WARNING:  Connection establishment time for session 63: 783 msecs
WARNING:  Connection establishment time for session 64: 686 msecs
WARNING:  Connection establishment time for session 86: 231 msecs
WARNING:  Connection establishment time for session 87: 231 msecs
WARNING:  Connection establishment time for session 88: 231 msecs
WARNING:  Connection establishment time for session 89: 302 msecs
WARNING:  Connection establishment time for session 94: 470 msecs
WARNING:  Connection establishment time for session 95: 462 msecs
WARNING:  Connection establishment time for session 96: 515 msecs
WARNING:  Connection establishment time for session 97: 462 msecs
WARNING:  Connection establishment time for session 98: 042 msecs
WARNING:  Connection establishment time for session 99: 786 msecs
WARNING:  Connection establishment time for session 100: 882 msecs
WARNING:  Connection establishment time for session 101: 755 msecs
WARNING:  Connection establishment time for session 102: 115 msecs
WARNING:  Connection establishment time for session 103: 873 msecs
WARNING:  Connection establishment time for session 104: 283 msecs
WARNING:  Connections to node 10.192.0.6:5432
WARNING:  Connection establishment time for session 47: 722 msecs
WARNING:  Connection establishment time for session 51: 750 msecs
WARNING:  Connection establishment time for session 52: 380 msecs
WARNING:  Connection establishment time for session 65: 473 msecs
WARNING:  Connection establishment time for session 66: 620 msecs
WARNING:  Connection establishment time for session 67: 680 msecs
WARNING:  Connection establishment time for session 90: 280 msecs
WARNING:  Connection establishment time for session 91: 280 msecs
WARNING:  Connection establishment time for session 92: 846 msecs
WARNING:  Connection establishment time for session 93: 350 msecs
WARNING:  Connection establishment time for session 105: 375 msecs
WARNING:  Connection establishment time for session 106: 283 msecs
WARNING:  Connection establishment time for session 107: 358 msecs
WARNING:  Connection establishment time for session 108: 297 msecs
WARNING:  Connection establishment time for session 109: 330 msecs
WARNING:  Connections to node 10.192.0.7:5432
WARNING:  Connection establishment time for session 48: 354 msecs
WARNING:  Connection establishment time for session 53: 139 msecs
WARNING:  Connection establishment time for session 54: 099 msecs
WARNING:  Connection establishment time for session 55: 887 msecs
WARNING:  Connection establishment time for session 56: 832 msecs
WARNING:  Connection establishment time for session 57: 836 msecs
WARNING:  Connection establishment time for session 58: 374 msecs
WARNING:  Connection establishment time for session 59: 170 msecs
WARNING:  Connection establishment time for session 60: 287 msecs
WARNING:  Connection establishment time for session 61: 378 msecs
WARNING:  Connection establishment time for session 68: 117 msecs
WARNING:  Connection establishment time for session 69: 038 msecs
WARNING:  Connection establishment time for session 70: 117 msecs
WARNING:  Connection establishment time for session 71: 109 msecs
WARNING:  Connection establishment time for session 72: 305 msecs
WARNING:  Connection establishment time for session 73: 238 msecs
WARNING:  Connection establishment time for session 74: 239 msecs
WARNING:  Connection establishment time for session 75: 239 msecs
WARNING:  Connection establishment time for session 76: 239 msecs
WARNING:  Connection establishment time for session 77: 283 msecs
WARNING:  Connection establishment time for session 78: 283 msecs
WARNING:  Connection establishment time for session 79: 644 msecs
WARNING:  Connection establishment time for session 80: 602 msecs
WARNING:  Connection establishment time for session 81: 116 msecs
WARNING:  Connection establishment time for session 82: 645 msecs
WARNING:  Connection establishment time for session 83: 093 msecs
WARNING:  Connection establishment time for session 84: 204 msecs
WARNING:  Connection establishment time for session 85: 233 msecs
  count   
----------
 42400100
(1 row)


select count(*) from test;
WARNING:  Connections to node 10.192.0.4:5432
WARNING:  Connection establishment time for session 110: 472 msecs
WARNING:  Connection establishment time for session 113: 913 msecs
WARNING:  Connection establishment time for session 114: 942 msecs
WARNING:  Connection establishment time for session 131: 704 msecs
WARNING:  Connection establishment time for session 132: 448 msecs
WARNING:  Connection establishment time for session 133: 767 msecs
WARNING:  Connection establishment time for session 150: 950 msecs
WARNING:  Connection establishment time for session 151: 674 msecs
WARNING:  Connection establishment time for session 152: 957 msecs
WARNING:  Connection establishment time for session 153: 025 msecs
WARNING:  Connection establishment time for session 171: 592 msecs
WARNING:  Connection establishment time for session 172: 697 msecs
WARNING:  Connection establishment time for session 173: 623 msecs
WARNING:  Connection establishment time for session 174: 595 msecs
WARNING:  Connection establishment time for session 175: 591 msecs
WARNING:  Connection establishment time for session 176: 140 msecs
WARNING:  Connection establishment time for session 177: 141 msecs
WARNING:  Connection establishment time for session 178: 385 msecs
WARNING:  Connection establishment time for session 179: 277 msecs
WARNING:  Connection establishment time for session 180: 277 msecs
WARNING:  Connection establishment time for session 181: 124 msecs
WARNING:  Connection establishment time for session 182: 054 msecs
WARNING:  Connection establishment time for session 183: 237 msecs
WARNING:  Connection establishment time for session 184: 215 msecs
WARNING:  Connection establishment time for session 185: 127 msecs
WARNING:  Connection establishment time for session 186: 220 msecs
WARNING:  Connection establishment time for session 187: 095 msecs
WARNING:  Connection establishment time for session 188: 123 msecs
WARNING:  Connection establishment time for session 194: 050 msecs
WARNING:  Connection establishment time for session 195: 059 msecs
WARNING:  Connection establishment time for session 196: 173 msecs
WARNING:  Connection establishment time for session 197: 065 msecs
WARNING:  Connection establishment time for session 198: 208 msecs
WARNING:  Connection establishment time for session 199: 050 msecs
WARNING:  Connection establishment time for session 200: 053 msecs
WARNING:  Connection establishment time for session 201: 087 msecs
WARNING:  Connection establishment time for session 202: 597 msecs
WARNING:  Connection establishment time for session 203: 056 msecs
WARNING:  Connection establishment time for session 204: 649 msecs
WARNING:  Connections to node 10.192.0.6:5432
WARNING:  Connection establishment time for session 111: 722 msecs
WARNING:  Connection establishment time for session 115: 117 msecs
WARNING:  Connection establishment time for session 116: 391 msecs
WARNING:  Connection establishment time for session 140: 234 msecs
WARNING:  Connection establishment time for session 141: 225 msecs
WARNING:  Connection establishment time for session 142: 248 msecs
WARNING:  Connection establishment time for session 167: 222 msecs
WARNING:  Connection establishment time for session 168: 267 msecs
WARNING:  Connection establishment time for session 169: 314 msecs
WARNING:  Connection establishment time for session 170: 185 msecs
WARNING:  Connection establishment time for session 189: 848 msecs
WARNING:  Connection establishment time for session 190: 205 msecs
WARNING:  Connection establishment time for session 191: 209 msecs
WARNING:  Connection establishment time for session 192: 205 msecs
WARNING:  Connection establishment time for session 193: 208 msecs
WARNING:  Connection establishment time for session 205: 979 msecs
WARNING:  Connection establishment time for session 206: 460 msecs
WARNING:  Connection establishment time for session 207: 460 msecs
WARNING:  Connection establishment time for session 208: 996 msecs
WARNING:  Connection establishment time for session 209: 002 msecs
WARNING:  Connection establishment time for session 210: 657 msecs
WARNING:  Connection establishment time for session 211: 721 msecs
WARNING:  Connection establishment time for session 212: 795 msecs
WARNING:  Connection establishment time for session 213: 767 msecs
WARNING:  Connection establishment time for session 214: 787 msecs
WARNING:  Connection establishment time for session 215: 729 msecs
WARNING:  Connection establishment time for session 216: 720 msecs
WARNING:  Connection establishment time for session 217: 737 msecs
WARNING:  Connection establishment time for session 218: 889 msecs
WARNING:  Connection establishment time for session 219: 608 msecs
WARNING:  Connections to node 10.192.0.7:5432
WARNING:  Connection establishment time for session 112: 354 msecs
WARNING:  Connection establishment time for session 117: 408 msecs
WARNING:  Connection establishment time for session 118: 375 msecs
WARNING:  Connection establishment time for session 119: 261 msecs
WARNING:  Connection establishment time for session 120: 264 msecs
WARNING:  Connection establishment time for session 121: 212 msecs
WARNING:  Connection establishment time for session 122: 199 msecs
WARNING:  Connection establishment time for session 123: 186 msecs
WARNING:  Connection establishment time for session 124: 172 msecs
WARNING:  Connection establishment time for session 125: 199 msecs
WARNING:  Connection establishment time for session 126: 475 msecs
WARNING:  Connection establishment time for session 127: 459 msecs
WARNING:  Connection establishment time for session 128: 306 msecs
WARNING:  Connection establishment time for session 129: 459 msecs
WARNING:  Connection establishment time for session 130: 454 msecs
WARNING:  Connection establishment time for session 134: 120 msecs
WARNING:  Connection establishment time for session 135: 142 msecs
WARNING:  Connection establishment time for session 136: 141 msecs
WARNING:  Connection establishment time for session 137: 142 msecs
WARNING:  Connection establishment time for session 138: 124 msecs
WARNING:  Connection establishment time for session 139: 140 msecs
WARNING:  Connection establishment time for session 143: 229 msecs
WARNING:  Connection establishment time for session 144: 077 msecs
WARNING:  Connection establishment time for session 145: 185 msecs
WARNING:  Connection establishment time for session 146: 059 msecs
WARNING:  Connection establishment time for session 147: 063 msecs
WARNING:  Connection establishment time for session 148: 368 msecs
WARNING:  Connection establishment time for session 149: 376 msecs
WARNING:  Connection establishment time for session 154: 285 msecs
WARNING:  Connection establishment time for session 155: 264 msecs
WARNING:  Connection establishment time for session 156: 081 msecs
WARNING:  Connection establishment time for session 157: 132 msecs
WARNING:  Connection establishment time for session 158: 081 msecs
WARNING:  Connection establishment time for session 159: 065 msecs
WARNING:  Connection establishment time for session 160: 068 msecs
WARNING:  Connection establishment time for session 161: 274 msecs
WARNING:  Connection establishment time for session 162: 091 msecs
WARNING:  Connection establishment time for session 163: 054 msecs
WARNING:  Connection establishment time for session 164: 098 msecs
WARNING:  Connection establishment time for session 165: 053 msecs
WARNING:  Connection establishment time for session 166: 092 msecs
  count   
----------
 42400100
(1 row)

Time: 21449.290 ms (00:21.449)
```

Or, task execution times of the same query on the almost same amount of data:
```
DEBUG:  Task execution time for task 1: 532 msecs
DEBUG:  Task execution time for task 33: 594 msecs
DEBUG:  Task execution time for task 2: 330 msecs
DEBUG:  Task execution time for task 36: 742 msecs
DEBUG:  Task execution time for task 39: 746 msecs
DEBUG:  Task execution time for task 60: 403 msecs
DEBUG:  Task execution time for task 42: 711 msecs
DEBUG:  Task execution time for task 87: 875 msecs
DEBUG:  Task execution time for task 57: 459 msecs
DEBUG:  Task execution time for task 78: 951 msecs
DEBUG:  Task execution time for task 54: 523 msecs
DEBUG:  Task execution time for task 99: 635 msecs
DEBUG:  Task execution time for task 51: 544 msecs
DEBUG:  Task execution time for task 48: 663 msecs
DEBUG:  Task execution time for task 45: 673 msecs
DEBUG:  Task execution time for task 96: 683 msecs
DEBUG:  Task execution time for task 72: 217 msecs
DEBUG:  Task execution time for task 84: 977 msecs
DEBUG:  Task execution time for task 93: 884 msecs
DEBUG:  Task execution time for task 75: 202 msecs
DEBUG:  Task execution time for task 123: 300 msecs
DEBUG:  Task execution time for task 114: 665 msecs
DEBUG:  Task execution time for task 111: 674 msecs
DEBUG:  Task execution time for task 66: 364 msecs
DEBUG:  Task execution time for task 102: 695 msecs
DEBUG:  Task execution time for task 90: 954 msecs
DEBUG:  Task execution time for task 63: 533 msecs
DEBUG:  Task execution time for task 69: 308 msecs
DEBUG:  Task execution time for task 81: 047 msecs
DEBUG:  Task execution time for task 105: 725 msecs
DEBUG:  Task execution time for task 117: 719 msecs
DEBUG:  Task execution time for task 108: 729 msecs
DEBUG:  Task execution time for task 120: 367 msecs
DEBUG:  Task execution time for task 126: 230 msecs
DEBUG:  Task execution time for task 3: 605 msecs
DEBUG:  Task execution time for task 4: 412 msecs
DEBUG:  Task execution time for task 35: 129 msecs
DEBUG:  Task execution time for task 38: 966 msecs
DEBUG:  Task execution time for task 5: 431 msecs
DEBUG:  Task execution time for task 41: 087 msecs
DEBUG:  Task execution time for task 62: 275 msecs
DEBUG:  Task execution time for task 44: 890 msecs
DEBUG:  Task execution time for task 6: 234 msecs
DEBUG:  Task execution time for task 47: 861 msecs
DEBUG:  Task execution time for task 56: 507 msecs
DEBUG:  Task execution time for task 50: 938 msecs
DEBUG:  Task execution time for task 7: 263 msecs
DEBUG:  Task execution time for task 8: 255 msecs
DEBUG:  Task execution time for task 9: 128 msecs
DEBUG:  Task execution time for task 113: 853 msecs
DEBUG:  Task execution time for task 65: 000 msecs
DEBUG:  Task execution time for task 80: 416 msecs
DEBUG:  Task execution time for task 95: 324 msecs
DEBUG:  Task execution time for task 107: 923 msecs
DEBUG:  Task execution time for task 53: 303 msecs
DEBUG:  Task execution time for task 10: 152 msecs
DEBUG:  Task execution time for task 83: 470 msecs
DEBUG:  Task execution time for task 110: 011 msecs
DEBUG:  Task execution time for task 86: 454 msecs
DEBUG:  Task execution time for task 77: 782 msecs
DEBUG:  Task execution time for task 74: 786 msecs
DEBUG:  Task execution time for task 101: 145 msecs
DEBUG:  Task execution time for task 71: 928 msecs
DEBUG:  Task execution time for task 119: 761 msecs
DEBUG:  Task execution time for task 11: 086 msecs
DEBUG:  Task execution time for task 89: 511 msecs
DEBUG:  Task execution time for task 92: 529 msecs
DEBUG:  Task execution time for task 98: 289 msecs
DEBUG:  Task execution time for task 12: 088 msecs
DEBUG:  Task execution time for task 13: 085 msecs
DEBUG:  Task execution time for task 59: 448 msecs
DEBUG:  Task execution time for task 122: 943 msecs
DEBUG:  Task execution time for task 68: 369 msecs
DEBUG:  Task execution time for task 14: 119 msecs
DEBUG:  Task execution time for task 104: 456 msecs
DEBUG:  Task execution time for task 125: 118 msecs
DEBUG:  Task execution time for task 128: 134 msecs
DEBUG:  Task execution time for task 116: 454 msecs
DEBUG:  Task execution time for task 15: 160 msecs
DEBUG:  Task execution time for task 16: 158 msecs
DEBUG:  Task execution time for task 17: 156 msecs
DEBUG:  Task execution time for task 18: 150 msecs
DEBUG:  Task execution time for task 19: 154 msecs
DEBUG:  Task execution time for task 20: 155 msecs
DEBUG:  Task execution time for task 21: 158 msecs
DEBUG:  Task execution time for task 34: 933 msecs
DEBUG:  Task execution time for task 22: 274 msecs
DEBUG:  Task execution time for task 23: 526 msecs
DEBUG:  Task execution time for task 26: 513 msecs
DEBUG:  Task execution time for task 28: 604 msecs
DEBUG:  Task execution time for task 24: 733 msecs
DEBUG:  Task execution time for task 25: 742 msecs
DEBUG:  Task execution time for task 27: 755 msecs
DEBUG:  Task execution time for task 29: 814 msecs
DEBUG:  Task execution time for task 32: 812 msecs
DEBUG:  Task execution time for task 31: 825 msecs
DEBUG:  Task execution time for task 30: 839 msecs
DEBUG:  Task execution time for task 49: 635 msecs
DEBUG:  Task execution time for task 40: 499 msecs
DEBUG:  Task execution time for task 46: 424 msecs
DEBUG:  Task execution time for task 37: 151 msecs
DEBUG:  Task execution time for task 43: 166 msecs
DEBUG:  Task execution time for task 58: 266 msecs
DEBUG:  Task execution time for task 76: 124 msecs
DEBUG:  Task execution time for task 82: 764 msecs
DEBUG:  Task execution time for task 79: 888 msecs
DEBUG:  Task execution time for task 55: 435 msecs
DEBUG:  Task execution time for task 52: 819 msecs
DEBUG:  Task execution time for task 61: 572 msecs
DEBUG:  Task execution time for task 88: 401 msecs
DEBUG:  Task execution time for task 64: 434 msecs
DEBUG:  Task execution time for task 97: 374 msecs
DEBUG:  Task execution time for task 73: 550 msecs
DEBUG:  Task execution time for task 91: 186 msecs
DEBUG:  Task execution time for task 106: 740 msecs
DEBUG:  Task execution time for task 94: 461 msecs
DEBUG:  Task execution time for task 118: 472 msecs
DEBUG:  Task execution time for task 85: 674 msecs
DEBUG:  Task execution time for task 103: 075 msecs
DEBUG:  Task execution time for task 67: 380 msecs
DEBUG:  Task execution time for task 109: 230 msecs
DEBUG:  Task execution time for task 112: 110 msecs
DEBUG:  Task execution time for task 70: 563 msecs
DEBUG:  Task execution time for task 100: 345 msecs
DEBUG:  Task execution time for task 121: 770 msecs
DEBUG:  Task execution time for task 127: 566 msecs
DEBUG:  Task execution time for task 115: 323 msecs
DEBUG:  Task execution time for task 124: 022 msecs
```

**Comment by onderkalaci** [comment]
This is not fixed yet

**Comment by onderkalaci** [comment]
fixed by #4923
